### ISSUE TITLE
allow diff comments for PRs

### ISSUE THREAD

**Issue by 2bndy5** [shenxianpeng/cpp-linter-action#11]
I've been digging through GH's REST API, and it looks like we can do an auto-review for PRs. I have to keep digging for more details, but I'd like to switch from posting a PR comment to creating a multitude of review comments (PR-synchronize events will have to discard stale reviews from prior workflow runs). This can also be done for pushed commits 😉 

Getting line number, file paths, and suggestions from clang-tidy & clang-format can be done, but it would take some extra dependencies. Clang-format can output suggested fixes in XML (only to stdout), and clang-tidy can output fix suggestions to a specified YAML file. Again, I need to investigate the best course of parsing these outputs from bash script.

**Comment by 2bndy5** [comment]
I'm starting to think that parsing the diff for files can actually be done with bash script (using a combination of `grep`, `tr`, & `sep` commands).

Latest dilemma is parsing a files diff that's split into "hunks". For example, this commits diff
![image](https://user-images.githubusercontent.com/14963867/130322460-f5ef412e-b366-4528-88e5-fa0c31514835.png)

looks like this in the JSON payload
```json
[{
  "files": [{
    "patch": "@@ -32,7 +32,7 @@ args=(\"$@\")\n FMT_STYLE=${args[0]}\n IFS=',' read -r -a FILE_EXT_LIST <<< \"${args[1]}\"\n TIDY_CHECKS=\"${args[2]}\"\n-# CLANG_VERSION=\"${args[3]}\"\n+CLANG_VERSION=\"${args[3]}\"\n \n echo \"GH_EVENT_PATH = $GITHUB_EVENT_PATH\"\n echo \"processing $GITHUB_EVENT_NAME event\"\n@@ -128,8 +128,8 @@ do\n       CLANG_CONFIG=\"--config\"\n    fi\n \n-   clang-tidy \"$filename\" \"$CLANG_CONFIG\" >> clang_tidy_report.txt\n-   clang-format -style=\"$FMT_STYLE\" --dry-run \"$filename\" >> clang_format_report.txt\n+   clang-tidy-\"$CLANG_VERSION\" \"$filename\" \"$CLANG_CONFIG\" >> clang_tidy_report.txt\n+   clang-format-\"$CLANG_VERSION\" -style=\"$FMT_STYLE\" --dry-run \"$filename\" >> clang_format_report.txt\n \n    echo \"Current Working Directory = $CWD\"\n    if [[ $(wc -l < clang_tidy_report.txt) -gt 0 ]]"
  }]
}]
```

**Comment by 2bndy5** [comment]
I've written some python scripts that will parse
1. the XML output from clang-format (uses python's builtin XML parser module)
2. the YML output from clang-tidy. This does not use the 3<sup>rd</sup> party `PyYaml` module as a dependency. However, I might have to switch to that dependency if I find that there are differences between YML outputs from different version of clang-tidy...

My intention is to keep using the _run_checks.sh_ script unless the a new user input called "diff-only" is asserted (set to 'true'). Unlike other actions that rely on python to parse clang-* outputs, I'm trying to avoid adding more dependencies and putting all the algorithms into a long and confusing singular python script.

## Should we switch to python completely?
I'm currently debating if we should completely switch over to python. This would replace the current dependencies `jq` & `curl` with python's 3<sup>rd</sup> party `requests` module (python has inbuilt `json` module that we can use to replace `jq`).

ps - I haven't committed anything yet since I'm obviously still brainstorming my approach.

**Comment by 2bndy5** [comment]
It looks like PyYAML is already installed in the ubuntu:latest image that the clang-tools container is built on. I should switch over...

**Comment by 2bndy5** [comment]
Been a while since I made progress on this issue, so here is an update:

Parsing diffs is proving more difficult than expected because of different scenarios presented in diffs vs how the REST API allocates comments in a diff.

I found out that github allows anyone to **suggest** changes in a PR's "Files Changed" tab.
A comment like this
~~~
this should be changed back when ready
```suggestion
      - uses: shenxianpeng/cpp-linter-action@master
```
~~~
will render like this
>[![image](https://user-images.githubusercontent.com/14963867/133342292-ac4d6b88-2bfc-481d-98c4-a65a87133cd1.png)](https://github.com/2bndy5/cpp-linter-action/pull/4/files?file-filters%5B%5D=.cpp&file-filters%5B%5D=.hpp&file-filters%5B%5D=.png&file-filters%5B%5D=.py&file-filters%5B%5D=.sh&file-filters%5B%5D=.xcf&file-filters%5B%5D=.yml#r703908228)

I'm not sure if BitBucket offers the same option (or similar). Notice the "commit suggestion" and "add suggestion to batch" buttons at the bottom of the rendered comment. These buttons allow the repo's maintainer(s) to directly commit changes from the comment(s) contents! 🥇 

There is 1 drawback though: The suggestion is limited to only the lines that are displayed in the diff. Some suggestions from clang-tidy involve multiple lines (like adding `{` and `}` or replacing `#pramga once` with proper header guards and a closing `#endif`). This makes outputting multi-line suggestions harder to stay scoped within the diffs' hunks. For simplicity with multi-line suggestions, I'll probably just output the clang-tidy notification (kinda like what is output currently) within the diff instead of trying to suggest a multi-line fix.

**Comment by shenxianpeng** [comment]
Thank you for the update. It looks complex, and it's OK if it can implement or not. 
I just tested Bitbucket, it also offers the same option. amazing!

![Uploading image.png…]()

**Comment by 2bndy5** [comment]
I've decided to only do diff comments for PRs because any diff comments made to a commit's diff will show in a PR as review comments. I'm starting to think twice about supporting push events for this reason (a "synchronized" event for a PR behaves identically to push events for a commit - except its for a different thread's URL)... We'll just have to see how it looks.

**Comment by 2bndy5** [comment]
BTW, there's a reason I have duplicate _.clang-format_ files (needs to be 1 for each _platformio.ini_ file)

**Comment by 2bndy5** [comment]
just a teaser
![image](https://user-images.githubusercontent.com/14963867/134293677-0f82ccd2-b00a-4dfd-872a-9638590e7d5c.png)

**Comment by 2bndy5** [comment]
I just found out that installing `clang-tools` comes with some handy extras like _/bin/clang-tidy-diff.py_ (which outputs clang-tidy warnings for a given unified diff) and _/bin/clang-format-diff.py_ (which can format changes in a given unified diff).I'm going to keep investigating, but using these extra tools requires a change to the docker image (installing `clang-tools-<version>` instead of `clang-format-<version>` and `clang-tidy-<version>` individually).

## Status
I have clang-tidy fixes and clang-format fixes aggregated and output as diff comments (with suggestions where available). However, I've found that doing this manually yields incomplete/imprecise suggestions in some situations.

example:
![image](https://user-images.githubusercontent.com/14963867/134446160-c2e7efce-88c1-456d-b665-8d9607e0200a.png)

The main problem is that the clang-tools outputs suggestions in terms of a file's offset (in bytes) instead of in terms of a file's line numbers.

# Conclusion
I'm going to roll back changes that include parsing clang-tools suggestions. I created a new `diff-only` input option (defaults to false - meaning action's original behavior) runs clang-tools against only what's in the event's diff - not the entire file(s) mentioned in the event's diff. I can keep the new `diff-only` user input option, but I'm going to roll back any changes related to posting comments within the event's diff. Maybe we can re-visit this issue later...

**Comment by 2bndy5** [comment]
I found out how code inspector is making annotations to each commit: it's using the [github checks API](https://docs.github.com/en/rest/reference/checks) (a subset of the REST API). 

I've started experimenting with this checks API on a different repo, but it looks like there's a required app-specific secret token involved... Additionally, it looks like we can post a max of 50 annotations per payload (via `requests` module), and we can do subsequent https requests if more than 50 annotations are needed. This will help avoid the REST API's usage limit (max of 25 comments).
